scalalib/src/mill/scalalib/JavaModule.scala

```diff
@@ -847,10 +847,61 @@ trait JavaModule
    */
   def generatedSources: T[Seq[PathRef]] = Task { Seq.empty[PathRef] }
 
+  /**
+   * A set of tasks producing generated code. The difference between this and
+   * [[generatedSources]] lies in the fact that IDE-related tasks
+   * (BSP/Bloop/GenIdea) will not run the tasks, but will use them to statically
+   * determine the location of the source code they will produce by looking up their
+   * [[Task#deferredGeneratedSourceRoots]], knowing that these roots will be present under
+   * the dest directories once the task is run.
+   */
+  def deferredGeneratedSourceTasks: Seq[NamedTask[_]] = Seq.empty
+
+  private def predictGeneratedSourceRoots(task: NamedTask[_])(workspace: os.Path): Seq[os.Path] = {
+    val dest = mill
+      .eval
+      .EvaluatorPaths
+      .resolveDestPaths(workspace / "out", task)
```

Comment:
Since the `out`-path is configurable, this must depend on the real out-path in some way.

Reply:
Yeah that's my concern, but I've not found an API for retrieving this bit of information

Reply:
Currently `Evaluator` has an `outPath` field, but it's not available in normal tasks like this one. Maybe we can expose a `out: os.Path` member on `mill.api.Ctx` for such advanced use cases, similar to how we already expose `workspace: os.Path`?

Reply:
Added an `out` member in [2d54b8f](https://github.com/com-lihaoyi/mill/pull/4621/commits/2d54b8fdab43e269ef4589471f01a945bce6e99a)